x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeSourceHandler.java

```diff
@@ -24,24 +24,55 @@
 /**
  * An {@link ExchangeSourceHandler} asynchronously fetches pages and status from multiple {@link RemoteSink}s
  * and feeds them to its {@link ExchangeSource}, which are created using the {@link #createExchangeSource()}) method.
- * {@link RemoteSink}s are added using the {@link #addRemoteSink(RemoteSink, int)}) method.
+ * {@link RemoteSink}s are added using the {@link #addRemoteSink(RemoteSink, int, ActionListener)}) method.
  *
  * @see #createExchangeSource()
- * @see #addRemoteSink(RemoteSink, int)
+ * @see #addRemoteSink(RemoteSink, int, ActionListener)
  */
 public final class ExchangeSourceHandler {
     private final ExchangeBuffer buffer;
     private final Executor fetchExecutor;
 
     private final PendingInstances outstandingSinks;
     private final PendingInstances outstandingSources;
-    private final FailureCollector failure = new FailureCollector();
+    private final boolean failFast;
+    private final FailureCollector failure;
 
-    public ExchangeSourceHandler(int maxBufferSize, Executor fetchExecutor) {
+    /**
+     * Creates a new ExchangeSourceHandler.
+     *
+     * @param maxBufferSize      the maximum size of the exchange buffer. A larger buffer reduces ``pauses`` but uses more memory,
+     *                           which could otherwise be allocated for other purposes.
+     * @param fetchExecutor      the executor used to fetch pages.
+     * @param failFast           whether this exchange source should stop fetching pages immediately when failures occur.
```

Comment:
I'm still not clear how this parameter is going to be useful. It's always set to true, and we can't change it according to skip_unavailable or any other settings, because those are per-cluster and `ExchangeSourceHandler` is common for the whole computation. I must be still missing some important part here.

Reply:
I added this flag to ExchangeSource based on the assumption that skip_unavailable should have been handled before reaching this path, except when dealing with allow_partial_results. Since we are discussing this topic, I’ve moved the flag to be per remote sink so that it can work with either outcome.